src/Interpreters/Context.cpp

```diff
@@ -5161,7 +5154,11 @@ std::shared_ptr<SessionLog> Context::getSessionLog() const
 
 std::shared_ptr<ZooKeeperLog> Context::getZooKeeperLog() const
 {
-    SharedLockGuard lock(shared->mutex);
+    std::lock_guard lock(mutex_shared_context);
+    if (!shared)
+        return {};
```

Comment:
This is one of the new addition for the race condition. And after discussing this with @mstetsyuk he pointed out that this shouldn't have any noticeable affect, because it will be only used at most once per ZooKeeper session. The same is true for `getAggregatedZooKeeperLog`.